### PR TITLE
[Fix] Fix SGD implementation

### LINKED ISSUE #440: [BUG] Zero momentum will disable weight decay in SGD optimizer implementation
**Describe the bug**
In mlx's SGD implementation:

```
    def apply_single(
        self, gradient: mx.array, parameter: mx.array, state: OptimizerState
    ):
        """Performs the SGD parameter update and stores :math:`v` in the
        optimizer state."""
        if self.momentum <= 0:
            return parameter - self.learning_rate * gradient

        v = state.get("v", mx.zeros_like(gradient))

        if self.weight_decay != 0:
            gradient += self.weight_decay * parameter

        v = self.momentum * v
        if self.dampening > 0:
            v += (1 - self.dampening) * gradient
        else:
            v += gradient

        if self.nesterov:
            update = gradient + self.momentum * v
        else:
            update = v
        state["v"] = v
        return parameter - self.learning_rate * update
```

If I don't set momentum, by default it is 0.0, so in 
```
        if self.momentum <= 0:
            return parameter - self.learning_rate * gradient
```
It will return immediately. If I set a weight decay without setting momentum, it essentially disables my weight decay because that line of code will not be reached.

From my understanding, momentum shouldn't conflict with weight decay, as in PyTorch's implementation: https://pytorch.org/docs/stable/generated/torch.optim.SGD.html#sgd, even if I don't set momentum, weight decay should still take effect.

**To Reproduce**
I'm learning this book "Dive into deep learning" and trying to implement its code using mlx, and I found this bug when implementing Linear regression
https://github.com/PRESIDENT810/d2l-en/blob/6cb8059e0ded2675871d7ae4d4ececfcd49ab25b/mlx_code/Chapter%203/3.7.ipynb

If I don't set momentum, as you can see the evaluation loss doesn't change much.
image

However, even if I set a very small positive momentum, evaluation decreases:
image

**Expected behavior**
Even if there is no momentum, weight decay still take effect and prevent model from overfitting.

**Desktop (please complete the following information):**
 - OS Version: [e.g. MacOS 14.1.2]
 - Version [e.g. 0.7.0]

**Additional context**
I think this should be an easy fix. Feel free to assign it to me.

### PR DESCRIPTION
**Pull request by PRESIDENT810**
## Proposed changes

This fixes #440 by correctly implementing SGD's weight decay and momentum. This changes includes
1. The original implementation will make weight decay not take any effect if not setting a momentum & set momentum <= 0. We should first adjust gradients regarding to weight decay then do momentum instead.
2. With dampening > 0, originally we always update `v` to `momentum * v + (1-dampening) * gradient`, but when we are in the first round of iteration where `v` is zero, we should update `v` to `gradient` instead of `(1-dampening) * gradient`.

Reference: I implemented SGD according to the algorithm described in https://pytorch.org/docs/stable/generated/torch.optim.SGD.html#sgd

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Review by angeloskath — CHANGES_REQUESTED** [review]
Hm good catch!

I think, however, that this implementation can be simplified. All we have to do is initialize `v` with `v = state.get("v", (self.dampening / self.momentum) * gradient)` and then it would be correct (including moving the momentum check as you did).

The 1st iteration will be doing some unnecessary work but I think that is completely fine. Also tbh I don't think it mattered that much that we were starting `v` with a slightly smaller gradient but still better to be correct.

**Comment by PRESIDENT810** [comment]
@angeloskath 
I tried to simplified it a little bit, but I still don't quite understand what do you mean by `v = state.get("v", (self.dampening / self.momentum) * gradient)`, especially why we have `self.dampening / self.momentum`?

Also I'm totally fine if you want me to change it back to use `(1 - self.dampening) * gradient` for the first iteration. I'm just following the algorithm displayed in that PyTorch link and don't know if that's going to make a big difference.

**Comment by angeloskath** [comment]
Oh I just meant that if you initialize `v` with `(self.dampening / self.momentum) * gradient` then you don't need the check for `first_iter`. Simply put `v` should be `gradient` in the first iteration, right? So

```python
v = state.get("v", (self.dampening / self.momentum) * gradient)
if self.dampening > 0:
    v = self.momentum * v + (1-self.dampening) * gradient
```
will result in

```python
v = self.momentum * (self.dampening / self.momentum) * gradient + (1-self.dampening) * gradient
  = gradient
```

**Comment by angeloskath** [comment]
In addition to the above the algorithm you implemented now is kinda wrong as it doesn't apply momentum at all when dampening is 0.

I would say the only things that needed changing were
1. moving the `if self.momentum <=0:` after the weight decay
2. initializing `v` with `(self.dampening / self.momentum) * gradient` so that the first iteration is exact

**Comment by PRESIDENT810** [comment]
Ah yes I see. I fixed it by moving momentum after weight decay and initializing `v` with default value `(self.dampening / self.momentum) * gradient` when dampening > 0. 
When dampening <= 0 I think we need to initialize v with default value 0, so we can have `v = gradient` for the first iteration.

Curious why a negative dampening value is treated as zero (although I know such a negative value doesn't make any sense).

**Comment by PRESIDENT810** [comment]
Ping

**Review by angeloskath — APPROVED** [review]
